File: FakturaCZ-Windows/app/services/pdf_renderer.py
"""PDF invoice renderer using ReportLab."""

from decimal import Decimal
from pathlib import Path

from reportlab.lib import colors
from reportlab.lib.pagesizes import A4
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
from reportlab.lib.units import mm
from reportlab.platypus import (
    SimpleDocTemplate, Table, TableStyle, Paragraph, Spacer, HRFlowable
)
from reportlab.lib.enums import TA_RIGHT, TA_LEFT

from app.models.invoice import Invoice
from app.services import persistence
from app.services.vat import czech_round, breakdown


def _fmt_currency(value: Decimal) -> str:
    rounded = czech_round(value)
    # Format with Czech locale style
    s = f"{rounded:,.2f}"
    s = s.replace(",", " ").replace(".", ",")
    return f"{s} Kč"


def _fmt_number(value: Decimal) -> str:
    return str(value)


def _fmt_date(iso_date: str) -> str:
    from datetime import date
    d = date.fromisoformat(iso_date)
    return d.strftime("%d.%m.%Y")


def render_invoice(invoice: Invoice, output_path: Path) -> Path:
    """Render an invoice to PDF and return the output path."""
    doc = SimpleDocTemplate(
        str(output_path),
        pagesize=A4,
        topMargin=20 * mm,
        bottomMargin=20 * mm,
        leftMargin=20 * mm,
        rightMargin=20 * mm,
    )

    styles = getSampleStyleSheet()
    styles.add(ParagraphStyle(
        "Title2", parent=styles["Title"], fontSize=20, spaceAfter=6
    ))
    styles.add(ParagraphStyle(
        "RightAligned", parent=styles["Normal"], alignment=TA_RIGHT
    ))
    styles.add(ParagraphStyle(
        "SmallGray", parent=styles["Normal"], fontSize=8, textColor=colors.gray
    ))
    styles.add(ParagraphStyle(
        "BoldRight", parent=styles["Normal"], alignment=TA_RIGHT,
        fontSize=14, fontName="Helvetica-Bold"
    ))

    story = []
    lines = invoice.get_lines()
    lines.sort(key=lambda l: l.sort_order)
    subject = persistence.get_subject(invoice.subject_registration_no or "")

    # -- Header --
    from app.models.enums import InvoiceType, InvoiceState
    type_label = InvoiceType[invoice.type].label.upper()
    state_label = InvoiceState[invoice.state].label

    story.append(Paragraph(f"{type_label} č. {invoice.invoice_number}", styles["Title2"]))
    story.append(Paragraph(f"Stav: {state_label}", styles["SmallGray"]))
    story.append(Spacer(1, 4 * mm))

    if invoice.is_locked:
        story.append(Paragraph(
            "🔒 Faktura je uzamčena pro účetní export",
            ParagraphStyle("Locked", parent=styles["Normal"],
                           fontSize=9, textColor=colors.HexColor("#D97706"))
        ))
        story.append(Spacer(1, 2 * mm))

    story.append(HRFlowable(width="100%", thickness=1, color=colors.gray))
    story.append(Spacer(1, 4 * mm))

    # -- Dates --
    date_data = [
        ["Datum vystavení:", _fmt_date(invoice.issued_date)],
        ["DUZP:", _fmt_date(invoice.taxable_supply_date)],
        ["Datum splatnosti:", _fmt_date(invoice.due_date)],
    ]
    date_table = Table(date_data, colWidths=[120, 200])
    date_table.setStyle(TableStyle([
        ("FONTSIZE", (0, 0), (-1, -1), 9),
        ("FONTNAME", (0, 0), (0, -1), "Helvetica-Bold"),
        ("BOTTOMPADDING", (0, 0), (-1, -1), 2),
        ("TOPPADDING", (0, 0), (-1, -1), 2),
    ]))
    story.append(date_table)
    story.append(Spacer(1, 4 * mm))

    # -- Subject --
    if subject:
        story.append(Paragraph("Odběratel:", styles["Heading4"]))
        story.append(Paragraph(subject.name, ParagraphStyle(
            "SubjectName", parent=styles["Normal"], fontName="Helvetica-Bold", fontSize=10
        )))
        story.append(Paragraph(subject.street, styles["Normal"]))
        story.append(Paragraph(f"{subject.zip} {subject.city}", styles["Normal"]))
        story.append(Paragraph(f"IČO: {subject.registration_no}", styles["Normal"]))
        if subject.vat_no:
            story.append(Paragraph(f"DIČ: {subject.vat_no}", styles["Normal"]))
        story.append(Spacer(1, 4 * mm))

    story.append(HRFlowable(width="100%", thickness=0.5, color=colors.gray))
    story.append(Spacer(1, 4 * mm))

    # -- Line Items Table --
    table_data = [["Popis", "Množství", "Jedn.", "Cena/ks", "DPH", "Celkem"]]
    for line in lines:
        table_data.append([
            line.item_description,
            _fmt_number(line.quantity),
            line.unit,
            _fmt_currency(line.unit_price),
            line.vat_rate_enum.label,
            _fmt_currency(line.line_total_with_vat),
        ])

    col_widths = [180, 50, 35, 80, 40, 80]
    line_table = Table(table_data, colWidths=col_widths)
    line_table.setStyle(TableStyle([
        ("FONTSIZE", (0, 0), (-1, -1), 9),
        ("FONTNAME", (0, 0), (-1, 0), "Helvetica-Bold"),
        ("BACKGROUND", (0, 0), (-1, 0), colors.HexColor("#F3F4F6")),
        ("GRID", (0, 0), (-1, -1), 0.5, colors.HexColor("#D1D5DB")),
        ("BOTTOMPADDING", (0, 0), (-1, -1), 4),
        ("TOPPADDING", (0, 0), (-1, -1), 4),
        ("ALIGN", (1, 0), (-1, -1), "RIGHT"),
        ("ALIGN", (0, 0), (0, -1), "LEFT"),
    ]))
    story.append(line_table)
    story.append(Spacer(1, 4 * mm))

    # -- VAT Breakdown --
    bkdn = breakdown(lines)
    if bkdn:
        vat_data = [["Sazba DPH", "Základ", "DPH"]]
        for rate in sorted(bkdn, key=lambda r: r.value, reverse=True):
            base, vat = bkdn[rate]
            vat_data.append([rate.label, _fmt_currency(base), _fmt_currency(vat)])

        vat_table = Table(vat_data, colWidths=[80, 100, 100])
        vat_table.setStyle(TableStyle([
            ("FONTSIZE", (0, 0), (-1, -1), 9),
            ("FONTNAME", (0, 0), (-1, 0), "Helvetica-Bold"),
            ("ALIGN", (1, 0), (-1, -1), "RIGHT"),
            ("BOTTOMPADDING", (0, 0), (-1, -1), 3),
        ]))
        story.append(vat_table)
        story.append(Spacer(1, 4 * mm))

    # -- Total --
    story.append(HRFlowable(width="100%", thickness=1, color=colors.black))
    story.append(Spacer(1, 2 * mm))
    total_data = [[
        Paragraph("Celkem k úhradě:", styles["Normal"]),
        Paragraph(_fmt_currency(invoice.total), styles["BoldRight"]),
    ]]
    total_table = Table(total_data, colWidths=[300, 165])
    story.append(total_table)
    story.append(Spacer(1, 4 * mm))

    # -- Payment Info --
    if invoice.variable_symbol or invoice.bank_account:
        story.append(Paragraph("Platební údaje:", styles["Heading4"]))
        if invoice.variable_symbol:
            story.append(Paragraph(f"Variabilní symbol: {invoice.variable_symbol}", styles["Normal"]))
        if invoice.bank_account:
            story.append(Paragraph(f"Číslo účtu: {invoice.bank_account}", styles["Normal"]))
        story.append(Spacer(1, 4 * mm))

    # -- Note --
    if invoice.note:
        story.append(Paragraph("Poznámka:", styles["Heading4"]))
        story.append(Paragraph(invoice.note, styles["Normal"]))

    doc.build(story)
    return output_path

File: FakturaCZ/FakturaCZ/Services/PDFRenderer.swift
import Foundation
import UIKit
import PDFKit

/// Renders an invoice to a PDF document using UIKit drawing.
enum PDFRenderer {

    static func render(invoice: Invoice) -> Data {
        let pageRect = CGRect(x: 0, y: 0, width: 595.2, height: 841.8)  // A4
        let margin: CGFloat = 50
        let contentWidth = pageRect.width - 2 * margin

        let renderer = UIGraphicsPDFRenderer(bounds: pageRect)

        return renderer.pdfData { context in
            context.beginPage()
            var y: CGFloat = margin

            // MARK: - Header
            let titleAttrs: [NSAttributedString.Key: Any] = [
                .font: UIFont.systemFont(ofSize: 22, weight: .bold),
                .foregroundColor: UIColor.black
            ]
            let title = invoice.type.label.uppercased() + " č. " + invoice.invoiceNumber
            title.draw(at: CGPoint(x: margin, y: y), withAttributes: titleAttrs)
            y += 35

            // MARK: - State badge
            let stateAttrs: [NSAttributedString.Key: Any] = [
                .font: UIFont.systemFont(ofSize: 11, weight: .medium),
                .foregroundColor: UIColor.darkGray
            ]
            let stateText = "Stav: \(invoice.state.label)"
            stateText.draw(at: CGPoint(x: margin, y: y), withAttributes: stateAttrs)
            y += 30

            // MARK: - Separator
            drawLine(in: context.cgContext, from: CGPoint(x: margin, y: y),
                     to: CGPoint(x: pageRect.width - margin, y: y))
            y += 15

            // MARK: - Dates
            let bodyAttrs: [NSAttributedString.Key: Any] = [
                .font: UIFont.systemFont(ofSize: 10),
                .foregroundColor: UIColor.black
            ]
            let labelAttrs: [NSAttributedString.Key: Any] = [
                .font: UIFont.systemFont(ofSize: 10, weight: .semibold),
                .foregroundColor: UIColor.black
            ]

            let dateFormatter = DateFormatter()
            dateFormatter.dateFormat = "dd.MM.yyyy"
            dateFormatter.locale = Locale(identifier: "cs_CZ")

            let dates: [(String, Date)] = [
                ("Datum vystavení:", invoice.issuedDate),
                ("Datum zdanitelného plnění:", invoice.taxableSupplyDate),
                ("Datum splatnosti:", invoice.dueDate)
            ]

            for (label, date) in dates {
                label.draw(at: CGPoint(x: margin, y: y), withAttributes: labelAttrs)
                let dateStr = dateFormatter.string(from: date)
                dateStr.draw(at: CGPoint(x: margin + 200, y: y), withAttributes: bodyAttrs)
                y += 16
            }
            y += 10

            // MARK: - Subject (buyer)
            if let subject = invoice.subject {
                "Odběratel:".draw(at: CGPoint(x: margin, y: y), withAttributes: labelAttrs)
                y += 18
                subject.name.draw(at: CGPoint(x: margin, y: y), withAttributes: bodyAttrs)
                y += 14
                subject.street.draw(at: CGPoint(x: margin, y: y), withAttributes: bodyAttrs)
                y += 14
                "\(subject.zip) \(subject.city)".draw(at: CGPoint(x: margin, y: y), withAttributes: bodyAttrs)
                y += 14
                "IČO: \(subject.registrationNo)".draw(at: CGPoint(x: margin, y: y), withAttributes: bodyAttrs)
                y += 14
                if let vatNo = subject.vatNo {
                    "DIČ: \(vatNo)".draw(at: CGPoint(x: margin, y: y), withAttributes: bodyAttrs)
                    y += 14
                }
                y += 10
            }

            // MARK: - Separator
            drawLine(in: context.cgContext, from: CGPoint(x: margin, y: y),
                     to: CGPoint(x: pageRect.width - margin, y: y))
            y += 15

            // MARK: - Line Items Table Header
            let headerAttrs: [NSAttributedString.Key: Any] = [
                .font: UIFont.systemFont(ofSize: 9, weight: .bold),
                .foregroundColor: UIColor.black
            ]
            let columns: [(String, CGFloat)] = [
                ("Popis", margin),
                ("Množství", margin + 230),
                ("Jedn.", margin + 300),
                ("Cena/ks", margin + 345),
                ("DPH", margin + 410),
                ("Celkem", margin + 450)
            ]
            for (header, x) in columns {
                header.draw(at: CGPoint(x: x, y: y), withAttributes: headerAttrs)
            }
            y += 16

            drawLine(in: context.cgContext, from: CGPoint(x: margin, y: y),
                     to: CGPoint(x: pageRect.width - margin, y: y), lineWidth: 0.5)
            y += 5

            // MARK: - Line Items
            let lineAttrs: [NSAttributedString.Key: Any] = [
                .font: UIFont.systemFont(ofSize: 9),
                .foregroundColor: UIColor.black
            ]

            let sortedLines = invoice.lines.sorted { $0.sortOrder < $1.sortOrder }
            for line in sortedLines {
                line.itemDescription.draw(at: CGPoint(x: margin, y: y), withAttributes: lineAttrs)
                formatNumber(line.quantity).draw(at: CGPoint(x: margin + 230, y: y), withAttributes: lineAttrs)
                line.unit.draw(at: CGPoint(x: margin + 300, y: y), withAttributes: lineAttrs)
                formatCurrency(line.unitPrice).draw(at: CGPoint(x: margin + 345, y: y), withAttributes: lineAttrs)
                line.vatRate.label.draw(at: CGPoint(x: margin + 410, y: y), withAttributes: lineAttrs)
                formatCurrency(line.lineTotalWithVAT).draw(at: CGPoint(x: margin + 450, y: y), withAttributes: lineAttrs)
                y += 15
            }

            y += 10
            drawLine(in: context.cgContext, from: CGPoint(x: margin, y: y),
                     to: CGPoint(x: pageRect.width - margin, y: y))
            y += 15

            // MARK: - VAT Breakdown
            let breakdown = VATCalculator.breakdown(for: sortedLines)
            for (rate, entry) in breakdown.sorted(by: { $0.key.rawValue > $1.key.rawValue }) {
                let breakdownText = "Základ \(rate.label): \(formatCurrency(entry.base))   DPH: \(formatCurrency(entry.vat))"
                breakdownText.draw(at: CGPoint(x: margin + 250, y: y), withAttributes: lineAttrs)
                y += 14
            }

            y += 10

            // MARK: - Totals
            let totalAttrs: [NSAttributedString.Key: Any] = [
                .font: UIFont.systemFont(ofSize: 14, weight: .bold),
                .foregroundColor: UIColor.black
            ]
            "Celkem k úhradě:".draw(at: CGPoint(x: margin + 250, y: y), withAttributes: labelAttrs)
            formatCurrency(invoice.total).draw(at: CGPoint(x: margin + 400, y: y), withAttributes: totalAttrs)
            y += 25

            // MARK: - Payment info
            if let vs = invoice.variableSymbol {
                "Variabilní symbol: \(vs)".draw(at: CGPoint(x: margin, y: y), withAttributes: bodyAttrs)
                y += 14
            }
            if let account = invoice.bankAccount {
                "Číslo účtu: \(account)".draw(at: CGPoint(x: margin, y: y), withAttributes: bodyAttrs)
                y += 14
            }

            // MARK: - Notes
            if let note = invoice.note, !note.isEmpty {
                y += 15
                "Poznámka:".draw(at: CGPoint(x: margin, y: y), withAttributes: labelAttrs)
                y += 16
                let noteRect = CGRect(x: margin, y: y, width: contentWidth, height: 60)
                note.draw(in: noteRect, withAttributes: bodyAttrs)
            }
        }
    }

    // MARK: - Drawing Helpers

    private static func drawLine(
        in cgContext: CGContext,
        from: CGPoint,
        to: CGPoint,
        lineWidth: CGFloat = 1.0
    ) {
        cgContext.setStrokeColor(UIColor.gray.cgColor)
        cgContext.setLineWidth(lineWidth)
        cgContext.move(to: from)
        cgContext.addLine(to: to)
        cgContext.strokePath()
    }

    private static func formatCurrency(_ value: Decimal) -> String {
        let number = NSDecimalNumber(decimal: CzechRounding.round(value))
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.minimumFractionDigits = 2
        formatter.maximumFractionDigits = 2
        formatter.groupingSeparator = " "
        formatter.decimalSeparator = ","
        formatter.locale = Locale(identifier: "cs_CZ")
        return (formatter.string(from: number) ?? number.stringValue) + " Kč"
    }

    private static func formatNumber(_ value: Decimal) -> String {
        let number = NSDecimalNumber(decimal: value)
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.minimumFractionDigits = 0
        formatter.maximumFractionDigits = 2
        formatter.locale = Locale(identifier: "cs_CZ")
        return formatter.string(from: number) ?? number.stringValue
    }
}
